### PR TITLE
allow extra webpack transform exclusions to be passed

### LINKED ISSUE #8: can't load react-globalize in vendor bundle
This is a bit involved, so please bear with me :)

In our app, we use default messages for calls to `messageFormatter`. These look like:

```js
const helloMsg = Globalize.messageFormatter('Hello {name}');
```

When react-globalize is imported in the module (or otherwise in the bundle), it patches `messageFormatter` to escape these strings, so the English message looks like:

```json
{
  "en": {
    "Hello (name)": "Hello {name}"
  }
}
```

When react-globalize is not loaded, the phrase is extracted as `"Hello {name}"`, which doesn't match the message and causes a runtime error.

To fix this, we tried putting `react-globalize` inside the vendor bundle, to ensure it is always loaded. This worked perfectly in development mode.

Unfortunately, in production, `globalize-webpack-plugin` rewrites all of the globalize module imports to refer to the i18n bundle, which is in a separate file. So with the following files:

```
vendor.js
i18n.en.js
main.js
```

react-globalize's imports are rewritten to refer to the patched globalize in `i18n.en.js`, not the original Globalize module in `vendor.js`, which fails at runtime.

I think the best solution would be for `globalize-webpack-plugin` to provide an exclusion mechanism for modules which should not have their module imports rewritten (similar to `util.isGlobalizeModule`). Then `react-globalize-webpack-plugin` could specify not to rewrite the react-globalize modules, and it could correctly live in vendor.js with the original Globalize import.

If this sounds good to you @rxaviers, I can get started on the patches (one here, one in react-globalize-webpack-plugin).

### PR DESCRIPTION
**Pull request by alunny**
fixes #8

Signed-off-by: Andrew Lunny <alunny@twitter.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +48,8 @@ new globalizePlugin({
 
 *output* is the name scheme of the built files.
 
+*transformExclusions* an array of functions to test on filepaths, and optionally exclude matching files from the dependency transformations. See [react-globalize-webpack-plugin](https://github.com/rxaviers/react-globalize-webpack-plugin) for an example usage.
+
```
Comment (rxaviers) [inline README.md]: Is react-globalize-webpack-plugin itself the example, or is there an example supposed to be there?

File: README.md [hunk 2]
```diff
@@ -33,7 +33,8 @@ new globalizePlugin({
 	developmentLocale: "en", // locale to be used for development.
 	supportedLocales: [ "en", "es", "zh", ... ], // locales that should be built support for.
 	messages: "messages/[locale].json", // messages (optional)
-	output: "globalize-compiled-data-[locale].[hash].js" // build output.
+	output: "globalize-compiled-data-[locale].[hash].js", // build output.
+	transformExclusions: [ myCoolFunction ]
```
Comment (rxaviers) [inline README.md]: Let's include a `// (optional)`

File: README.md (hunk 1 above)
Comment (rxaviers) [inline README.md]: It's important to mention this only takes effect on Production environment.

Comment (rxaviers) [inline README.md]: Thinking loud... It's not only excluded from the dependency transformations, but also from extracting the formatters and everything else.

File: ProductionModePlugin.js [hunk 3]
```diff
@@ -65,6 +73,7 @@ ProductionModePlugin.prototype.apply = function(compiler) {
     var request = this.state.current.request;
     if(param.isString() && param.string === "globalize" &&
           !util.isGlobalizeModule(request) &&
+          !isTransformExcluded(request) &&
           !(globalizeCompilerHelper.isCompiledDataModule(request))) {
```
Comment (rxaviers) [inline ProductionModePlugin.js]: Can we simplify this and do something like the below (where returning `true` means process it, `false` skip it)?

```diff
-          !util.isGlobalizeModule(request) &&
-          !isTransformExcluded(request) &&
-          !(globalizeCompilerHelper.isCompiledDataModule(request))) {
+         moduleFilter(request))) {
```

Comment (rxaviers) [inline ProductionModePlugin.js]: For clarity, I am wondering if we can leverage your change to clean up this former mess.

**Comment by rxaviers** [comment]
What do you think of the following proposed changes?

1. Use `transformExclusions` in development mode too, because there's no prejudice doing so and its usage become more consistent in both modes.
1. Rename the attribute... (a) I'd like to include `filter` in its name, because the function is supposed to return a boolean similar to what [Array.prototype.filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) does and associating the name (with a known API) helps deducing that; (b) I'd like to make the name more generic, it seems like it doesn't only exclude it from transforming the require, but excludes from everything else (the module gets pretty much skipped from globalize plugin processings); A name suggestion: `moduleFilter` (returning `true` means process it, returning `false` means skip it). Open to other suggestions.
1. Accept a function rather than an Array of functions. The various logics could be handled inside the function. This is a consequence of my proposed change above from `transformExclusions` (plural) to `moduleFilter` (singular).

File: README.md (hunk 1 above)
Comment (rxaviers) [inline README.md]: Follow up comment in https://github.com/rxaviers/globalize-webpack-plugin/pull/9#issuecomment-157363068

**Comment by alunny** [comment]
I like all of those suggestions. I'll update the change accordingly.

**Comment by rxaviers** [comment]
I see you have pushed new changes. Note there's no notifications for commit pushes. Please, is it ready for review? Thanks

**Comment by alunny** [comment]
Sorry, haven't had a chance to look at this in the last couple of weeks. Will follow up soon :)

**Comment by alunny** [comment]
Hi @rxaviers - sorry for the long delay, but yes this is ready for review.

File: util.js [hunk 4]
```diff
@@ -33,6 +35,18 @@ module.exports = {
       (j !== -1 /* 3 */ && filepath.length - j === 1 /* 4 */);
   },
 
+  moduleFilterFn: function(filterFn) {
```
Comment (rxaviers) [inline util.js]: `s/filterFn/moduleFilter`?

**Comment by rxaviers** [comment]
Other than the one comment, it seems good to me.

File: ProductionModePlugin.js [hunk 5]
```diff
@@ -63,8 +65,7 @@ ProductionModePlugin.prototype.apply = function(compiler) {
   // Globalize object.
   compiler.parser.plugin("call require:commonjs:item", function(expr, param) {
     var request = this.state.current.request;
-    if(param.isString() && param.string === "globalize" &&
-          !util.isGlobalizeModule(request) &&
+    if(param.isString() && param.string === "globalize" && moduleFilter(filepath) &&
```
Comment (rxaviers) [inline ProductionModePlugin.js]: s/moduleFilter(filepath)/moduleFilter(request)?

**Comment by rxaviers** [comment]
We definitely need some unit tests for this project. :see_no_evil:

**Comment by rxaviers** [comment]
I'm amending these two small fixes myself...

**Comment by rxaviers** [comment]
Closed by 9211a2af31dbb5825b41e7037bc3149b06a7a111

**Comment by alunny** [comment]
Thanks so much @rxaviers :)

**Comment by rxaviers** [comment]
Thank you.

**Pull request closed without merging**